tools/autograd/derivatives.yaml

```diff
@@ -1775,6 +1775,19 @@
   target: smooth_l1_loss_backward(grad, target, self, reduction, beta)
   result: apply_loss_reduction(smooth_l1_loss_backward(self_t.conj(), self_p, target_p, at::Reduction::None, beta).conj() + smooth_l1_loss_backward(target_t.conj(), target_p, self_p, at::Reduction::None, beta).conj(), reduction)
 
+- name: margin_ranking_loss(Tensor input1, Tensor input2, Tensor target, float margin=0.0, int reduction=Mean) -> Tensor
+  input1, input2: "grad.defined() ? sub_backward(margin_ranking_loss_backward_input1(grad, input1, input2, target, margin, reduction)) : std::tuple<Tensor, Tensor>()"
```

Comment:
Do you really need input1 and input2 here?
I feel like making this op non-composite is making a lot of things more complex than they need to be?

Reply:
You need it to recompute the threshold. This could potentially be avoided by computing the threshold in the forward and passing it to the backward functions. It's not clear whether we'll try and benchmark that other approach or not.

This was some sort of introductory PR to TI / derivatives / benchmarking and so on for @nkaretnikov. Benchmarks (https://github.com/pytorch/pytorch/pull/72327#issuecomment-1029774569) show that we managed to make it faster, so there was progress in that direction. Now, it is indeed the case that the code here is much more complex than the previous one, but we got a x2 speed-up, so I guess that's fine :)

Reply:
Much better speed is definitely a good argument! Note that memory usage is going to be a lot larger though (given all the buffers you save for backward).
I'm happy with the change if we say the speed benefit is worth it!

Reply:
Before it was a compositeimplicit op. So not sure what it was saving without checking the code in details.
But if we actually did `y * (input1 - input2)` then we would only save 1 or 2 Tensor of the input size (result of - and y). While the new function saves all 3 now.